staging/src/k8s.io/cri-api/README.md

```diff
@@ -6,9 +6,194 @@ without the need to recompile. CRI consists of a protocol buffers and gRPC API.
+Please do **not** file issues or submit PRs against the [kubernetes/cri-api](https://github.com/kubernetes/cri-api)
 repository as it is readonly, all development is done in [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes).
 
+The CRI API is defined in [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes)
+repository and is **only** intended to be used for kubelet to container runtime 
+interactions, or for node-level troubleshooting using a tool such as `crictl`.
+It is **not** a common purpose container runtime API for general use, and is intended
+to be Kubernetes-centric. We try to avoid it, but there may be logic within a container
+runtime that optimizes for the order or specific parameters of call(s) that the kubelet
+makes.
+
+## Version skew policy
+
+On a single Node there may be installed multiple components implementing
+different versions of CRI API.
+
+For example, on a single node there might be:
+
+- _Kubelet_ may call into _Container Runtime_ (e.g. [containerd](https://containerd.io))
+  and _Image Service Proxy_ (e.g. [stargz-snapshotter](https://github.com/containerd/stargz-snapshotter)).
+  _Container Runtime_ may be versioned with the OS Image, _Kubelet_ is installed
+  by system administrator and _Image Service proxy_ is versioned by the third party vendor.
+- _Image Service Proxy_ calls into _Container Runtime_.
+- _CRI tools_ (e.g. [crictl](https://kubernetes.io/docs/tasks/debug/debug-cluster/crictl/))
+  may be installed by end user to troubleshoot, same as a third party daemonsets.
+  All of them are used to call into the _Container Runtime_ to collect container information.
+
+So on a single node it may happen that _Container Runtime_ is serving a newer
+version'd kubelet and older versioned crictl. This is a supported scenario within
+the version skew policy.
+
+### Version Skew Policy for CRI API
+
+CRI API has two versions:
+- Major semantic version (known versions are `v1alpha2` ([removed in 1.26](https://kubernetes.io/blog/2022/12/09/kubernetes-v1-26-release/#cri-v1alpha2-removed)), `v1`).
+- Kubernetes version (for example: `@1.23`). Note, the `cri-api` Golang library is versioned as `0.23` as it doesn't guarantee Go types backward compatibility.
+
+Major semantic version (e.g. `v1`) is used to introduce breaking changes
+and major new features that are incompatible with the current API.
+
+Kubernetes version is used to indicate a specific feature set implemented
+on top of the major semantic version. All changes made without the change
+of a major semantic version API must be backward and forward compatible.
+
+- _Kubelet_ must work with the older _Container Runtime_ if it implements
+  the same semantic version of CRI API (e.g. `v1`) of up to three Kubernetes minor
+  versions back. New features implemented in CRI API must be gracefully degraded.
+  For example, _Kubelet_ of version 1.26 must work with _Container Runtime_
+  implementing `k8s.io/cri-api@v0.23.0`+.
```

Comment:
I think we must flesh out what it means in terms of proto changes that are allowed to support any skew. Also, examples of how things will behave when kubelet and runtime aren't on the same version will be helpful to make a call here.

Reply:
Thanks @SergeyKanzhelev, this document is really much appreciated 👍

I too would like to understand a bit better what kind of breaking changes would be allowed inside a stable major semantic version (`v1`). But OTOH I think we can merge this in it's current form and making further improvements in separate PRs ->
/lgtm